Dockerfile: update to Debian "bookworm", and update frozen images to debian:bookworm-slim

thaJeztah opened the pull request:
- [x] depends on https://github.com/moby/moby/pull/46852
- [x] depends on https://github.com/moby/moby/pull/46854
- [x] depends on https://github.com/moby/moby/pull/46859


### quota: increase sparse test-image to 300MB
Starting with [6e0ed3d19c54603f0f7d628ea04b550151d8a262], the minimum
allowed size is now 300MB. Given that this is a sparse image, and
the size of the image is irrelevant to the test (we check for
limits defined through project-quotas, not the size of the
device itself), we can raise the size of this image.

[6e0ed3d19c54603f0f7d628ea04b550151d8a262]: https://git.kernel.org/pub/scm/fs/xfs/xfsprogs-dev.git/commit/?id=6e0ed3d19c54603f0f7d628ea04b550151d8a262


### Dockerfile: update to Debian "bookworm" (current stable)

Debian Woodworm ships with a newer version of iptables, which caused two
tests to fail:

    === FAIL: amd64.integration-cli TestDockerDaemonSuite/TestDaemonICCLinkExpose (1.18s)
    docker_cli_daemon_test.go:841: assertion failed: false (matched bool) != true (true bool): iptables output should have contained "DROP.*all.*ext-bridge6.*ext-bridge6", but was "Chain FORWARD (policy ACCEPT 0 packets, 0 bytes)\n pkts bytes target     prot opt in     out     source               destination         \n    0     0 DOCKER-USER  0    --  *      *       0.0.0.0/0            0.0.0.0/0           \n    0     0 DOCKER-ISOLATION-STAGE-1  0    --  *      *       0.0.0.0/0            0.0.0.0/0           \n    0     0 ACCEPT     0    --  *      ext-bridge6  0.0.0.0/0            0.0.0.0/0            ctstate RELATED,ESTABLISHED\n    0     0 DOCKER     0    --  *      ext-bridge6  0.0.0.0/0            0.0.0.0/0           \n    0     0 ACCEPT     0    --  ext-bridge6 !ext-bridge6  0.0.0.0/0            0.0.0.0/0           \n    0     0 DROP       0    --  ext-bridge6 ext-bridge6  0.0.0.0/0            0.0.0.0/0           \n"
    --- FAIL: TestDockerDaemonSuite/TestDaemonICCLinkExpose (1.18s)

    === FAIL: amd64.integration-cli TestDockerDaemonSuite/TestDaemonICCPing (1.19s)
    docker_cli_daemon_test.go:803: assertion failed: false (matched bool) != true (true bool): iptables output should have contained "DROP.*all.*ext-bridge5.*ext-bridge5", but was "Chain FORWARD (policy ACCEPT 0 packets, 0 bytes)\n pkts bytes target     prot opt in     out     source               destination         \n    0     0 DOCKER-USER  0    --  *      *       0.0.0.0/0            0.0.0.0/0           \n    0     0 DOCKER-ISOLATION-STAGE-1  0    --  *      *       0.0.0.0/0            0.0.0.0/0           \n    0     0 ACCEPT     0    --  *      ext-bridge5  0.0.0.0/0            0.0.0.0/0            ctstate RELATED,ESTABLISHED\n    0     0 DOCKER     0    --  *      ext-bridge5  0.0.0.0/0            0.0.0.0/0           \n    0     0 ACCEPT     0    --  ext-bridge5 !ext-bridge5  0.0.0.0/0            0.0.0.0/0           \n    0     0 DROP       0    --  ext-bridge5 ext-bridge5  0.0.0.0/0            0.0.0.0/0           \n"
    --- FAIL: TestDockerDaemonSuite/TestDaemonICCPing (1.19s)

Both the `TestDaemonICCPing`, and `TestDaemonICCLinkExpose` test were introduced
in dd0666e64f17329355c77aae1a2ac0fe2fe43402. These tests called `iptables` with
the `-n` option. This option prevents it from doing a reverse-DNS lookup as an
optimization.

However, the `-n` option did not have an effect to the `prot` column before
commit [da8ecc62dd765b15df84c3aa6b83dcb7a81d4ffa] (iptables < v1.8.9 or v1.8.8).
Newer versions, such as the iptables version shipping with Debian Woodworm do,
so we need to update the expected output for this version.

With this change;

    make DOCKER_GRAPHDRIVER=vfs TEST_FILTER=TestDaemonICC  TEST_IGNORE_CGROUP_CHECK=1  test-integration
    ...
    --- PASS: TestDockerDaemonSuite (139.11s)
        --- PASS: TestDockerDaemonSuite/TestDaemonICCLinkExpose (54.62s)
        --- PASS: TestDockerDaemonSuite/TestDaemonICCPing (84.48s)

[da8ecc62dd765b15df84c3aa6b83dcb7a81d4ffa]: https://git.netfilter.org/iptables/commit/?id=da8ecc62dd765b15df84c3aa6b83dcb7a81d4ffa


### Dockerfile: remove uses of DEBIAN_FRONTEND

We used DEBIAN_FRONTEND in some places to prevent installation of packages
from being blocked. However, debian bookworm now [includes a fix][1] for
situations like this (it was specifically reported for Docker situations <3),
so we can get rid of these.

Thanks to Tianon for noticing this, and for linking to the Debian ticket!

[1]: https://bugs.debian.org/929417


### frozen images: update to debian:bookworm-slim

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

**- How I did it**

**- How to verify it**

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

thaJeztah committed:
Dockerfile: update to Debian "bookworm" (current stable)

Also switch yamllint to be installed from debian's packages, which are
currently at v1.29.0.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
frozen images: update to debian:bookworm-slim

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
Hmm... looks like libseccomp issues; on arm, ppc64le and s390x

```
 > [runc-build 3/3] RUN --mount=from=runc-src,src=/usr/src/runc,rw     --mount=type=cache,target=/root/.cache/go-build,id=runc-build-linux/arm/v6 <<EOT (set -e...):
0.492 CGO_ENABLED=1 go build -trimpath  -tags "seccomp netgo osusergo" -ldflags "-extldflags -static -X main.gitCommit=v1.1.7-0-g860f061 -X main.version=1.1.7 " -o runc .
59.14 go build github.com/opencontainers/runc/vendor/github.com/seccomp/libseccomp-golang:
59.14 # pkg-config --cflags  -- libseccomp
59.14 Package libseccomp was not found in the pkg-config search path.
59.14 Perhaps you should add the directory containing `libseccomp.pc'
59.14 to the PKG_CONFIG_PATH environment variable
59.14 Package 'libseccomp', required by 'virtual:world', not found
59.14 pkg-config: exit status 1
69.39 make: *** [Makefile:41: static] Error 1
```

thaJeztah commented:
yamllint installation is also unhappy;

```
#114 [dev-base 10/13] RUN pip3 install yamllint==1.27.1
#114 0.355 error: externally-managed-environment
#114 0.355 
#114 0.355 × This environment is externally managed
#114 0.355 ╰─> To install Python packages system-wide, try apt install
#114 0.355     python3-xyz, where xyz is the package you are trying to
#114 0.355     install.
#114 0.355     
#114 0.355     If you wish to install a non-Debian-packaged Python package,
#114 0.355     create a virtual environment using python3 -m venv path/to/venv.
#114 0.355     Then use path/to/venv/bin/python and path/to/venv/bin/pip. Make
#114 0.355     sure you have python3-full installed.
#114 0.355     
#114 0.355     If you wish to install a non-Debian packaged Python application,
#114 0.355     it may be easiest to use pipx install xyz, which will manage a
#114 0.355     virtual environment for you. Make sure you have pipx installed.
#114 0.355     
#114 0.355     See /usr/share/doc/python3.11/README.venv for more information.
#114 0.355 
#114 0.355 note: If you believe this is a mistake, please contact your Python installation or OS distribution provider. You can override this, at the risk of breaking your Python installation or OS, by passing --break-system-packages.
#114 0.355 hint: See PEP 668 for the detailed specification.
#114 ERROR: process "/bin/sh -c pip3 install yamllint==${YAMLLINT_VERSION}" did not complete successfully: exit code: 1
------
 > [dev-base 10/13] RUN pip3 install yamllint==1.27.1:
0.355     sure you have python3-full installed.
0.355     
0.355     If you wish to install a non-Debian packaged Python application,
0.355     it may be easiest to use pipx install xyz, which will manage a
0.355     virtual environment for you. Make sure you have pipx installed.
0.355     
0.355     See /usr/share/doc/python3.11/README.venv for more information.
0.355 
0.355 note: If you believe this is a mistake, please contact your Python installation or OS distribution provider. You can override this, at the risk of breaking your Python installation or OS, by passing --break-system-packages.
0.355 hint: See PEP 668 for the detailed specification.
```

tianon commented on Dockerfile:
```diff
@@ -33,8 +33,8 @@ FROM --platform=$BUILDPLATFORM ${GOLANG_IMAGE} AS base
 RUN echo 'Binary::apt::APT::Keep-Downloaded-Packages "true";' > /etc/apt/apt.conf.d/keep-cache
 ARG APT_MIRROR
-RUN sed -ri "s/(httpredir|deb).debian.org/${APT_MIRROR:-deb.debian.org}/g" /etc/apt/sources.list \
- && sed -ri "s/(security).debian.org/${APT_MIRROR:-security.debian.org}/g" /etc/apt/sources.list
+RUN sed -ri "s/(httpredir|deb).debian.org/${APT_MIRROR:-deb.debian.org}/g" /etc/apt/sources.list.d/debian.sources \
+ && sed -ri "s/(security).debian.org/${APT_MIRROR:-security.debian.org}/g" /etc/apt/sources.list.d/debian.sources
```
Should we just axe this instead? :eyes:

tianon commented on Dockerfile:
```diff
@@ -33,8 +33,8 @@ FROM --platform=$BUILDPLATFORM ${GOLANG_IMAGE} AS base
- && sed -ri "s/(security).debian.org/${APT_MIRROR:-security.debian.org}/g" /etc/apt/sources.list
+RUN sed -ri "s/(httpredir|deb).debian.org/${APT_MIRROR:-deb.debian.org}/g" /etc/apt/sources.list.d/debian.sources \
+ && sed -ri "s/(security).debian.org/${APT_MIRROR:-security.debian.org}/g" /etc/apt/sources.list.d/debian.sources
 ARG DEBIAN_FRONTEND
```
Given the fix for https://bugs.debian.org/929417 made it into bookworm, we _should_ be able to remove all remaining instances of `DEBIAN_FRONTEND` (generally, everywhere) too :eyes:

cpuguy83 replied:
WHHHAAAAA?! That's exciting!

tianon commented:
> Hmm... looks like libseccomp issues; on arm, ppc64le and s390x

I _really_ hope I'm right, because this is going to be the most _ironic_ case of the "bookworm causes seccomp profile failures" I've seen yet -- which version of Docker, runc, and libseccomp2 are on those three hosts?  Does the build work if you disable seccomp or update Docker and/or libseccomp2?  :eyes:

(I can't reproduce on my machine with QEMU on s390x :see_no_evil:)

crazy-max commented:
Since bookworm, pkgconf cross pkg needs to be installed. More info https://github.com/tonistiigi/xx/pull/111

thaJeztah committed:
Dockerfile: remove uses of DEBIAN_FRONTEND

We used DEBIAN_FRONTEND in some places to prevent installation of packages
from being blocked. However, debian bookworm now [includes a fix][1] for
situations like this (it was specifically reported for Docker situations <3),
so we can get rid of these.

Thanks to Tianon for noticing this, and for linking to the Debian ticket!

[1]: https://bugs.debian.org/929417

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

crazy-max commented on quota/testhelpers.go:
```diff
@@ -28,6 +26,12 @@ func CanTestQuota() (string, bool) {
+	// is 300MB.
+	//
+	// See https://git.kernel.org/pub/scm/fs/xfs/xfsprogs-dev.git/commit/?id=6e0ed3d19c54603f0f7d628ea04b550151d8a262
+	const imageSize = 64 * 1024 * 1024
```
Should be `300 * 1024 * 1024` no?

thaJeztah replied:
Ah, crap 🙈

thaJeztah replied:
Fixed 😂

thaJeztah commented:
Failures with integration tests with `systemd` (both 20.04 and 22.04);

```
---> Making bundle: .integration-daemon-start (in bundles/test-integration)
Using test binary /usr/local/cli-integration/docker
# DOCKER_EXPERIMENTAL is set: starting daemon with experimental features enabled! 
Starting apparmor (via systemctl): apparmor.serviceAssertion failed on job for apparmor.service.
 failed!
```

thaJeztah commented:
Some other failures;

```bash
=== Failed
=== FAIL: amd64.integration-cli TestDockerDaemonSuite/TestDaemonICCLinkExpose (1.19s)
    docker_cli_daemon_test.go:841: assertion failed: false (matched bool) != true (true bool): iptables output should have contained "DROP.*all.*ext-bridge6.*ext-bridge6", but was "Chain FORWARD (policy ACCEPT 0 packets, 0 bytes)\n pkts bytes target     prot opt in     out     source               destination         \n    0     0 DOCKER-USER  0    --  *      *       0.0.0.0/0            0.0.0.0/0           \n    0     0 DOCKER-ISOLATION-STAGE-1  0    --  *      *       0.0.0.0/0            0.0.0.0/0           \n    0     0 ACCEPT     0    --  *      ext-bridge6  0.0.0.0/0            0.0.0.0/0            ctstate RELATED,ESTABLISHED\n    0     0 DOCKER     0    --  *      ext-bridge6  0.0.0.0/0            0.0.0.0/0           \n    0     0 ACCEPT     0    --  ext-bridge6 !ext-bridge6  0.0.0.0/0            0.0.0.0/0           \n    0     0 DROP       0    --  ext-bridge6 ext-bridge6  0.0.0.0/0            0.0.0.0/0           \n"
    --- FAIL: TestDockerDaemonSuite/TestDaemonICCLinkExpose (1.19s)

=== FAIL: amd64.integration-cli TestDockerDaemonSuite/TestDaemonICCPing (1.21s)
    docker_cli_daemon_test.go:803: assertion failed: false (matched bool) != true (true bool): iptables output should have contained "DROP.*all.*ext-bridge5.*ext-bridge5", but was "Chain FORWARD (policy ACCEPT 0 packets, 0 bytes)\n pkts bytes target     prot opt in     out     source               destination         \n    0     0 DOCKER-USER  0    --  *      *       0.0.0.0/0            0.0.0.0/0           \n    0     0 DOCKER-ISOLATION-STAGE-1  0    --  *      *       0.0.0.0/0            0.0.0.0/0           \n    0     0 ACCEPT     0    --  *      ext-bridge5  0.0.0.0/0            0.0.0.0/0            ctstate RELATED,ESTABLISHED\n    0     0 DOCKER     0    --  *      ext-bridge5  0.0.0.0/0            0.0.0.0/0           \n    0     0 ACCEPT     0    --  ext-bridge5 !ext-bridge5  0.0.0.0/0            0.0.0.0/0           \n    0     0 DROP       0    --  ext-bridge5 ext-bridge5  0.0.0.0/0            0.0.0.0/0           \n"
    --- FAIL: TestDockerDaemonSuite/TestDaemonICCPing (1.21s)
```

iptables output should have contained `"DROP.*all.*ext-bridge6.*ext-bridge6"`
Looks like output shows;

```
DROP       0    --  ext-bridge6 ext-bridge6  0.0.0.0/0            0.0.0.0/0
```

thaJeztah commented:
For the apparmor issue, looking at the failure again, I guess `apparmor.serviceAssertion` was actually not a single word, but it's 2 lines printed together, so should've been;

```
Starting apparmor (via systemctl): apparmor.service
Assertion failed on job for apparmor.service.
 failed!
```

I noticed that we're manually starting apparmor using this script https://github.com/moby/moby/blob/ce1ee98aba41e117c59bf17a108191eed14e9b9c/hack/make/.integration-daemon-start#L94-L103, and was a bit confused why `systemctl` showed up in the failure, but it looks like that script prints `via systemctl`;

```bash
/etc/init.d/apparmor start
[ ok ] Starting apparmor (via systemctl): apparmor.service.
```

And for sure, looking on an Ubuntu machine, there's indeed a systemd unit as well, which looks to be a wrapper for the `/etc/init.d/apparmor` scripts, so it's turtles all the way down;

```bash
systemctl cat apparmor
# /lib/systemd/system/apparmor.service
[Unit]
Description=AppArmor initialization
After=local-fs.target
Before=sysinit.target
AssertPathIsReadWrite=/sys/kernel/security/apparmor/.load
ConditionSecurity=apparmor
DefaultDependencies=no
Documentation=man:apparmor(7)
Documentation=http://wiki.apparmor.net/

# Don't start this unit on the Ubuntu Live CD
ConditionPathExists=!/rofs/etc/apparmor.d

[Service]
Type=oneshot
RemainAfterExit=yes
ExecStart=/etc/init.d/apparmor start
ExecStop=/etc/init.d/apparmor stop
ExecReload=/etc/init.d/apparmor reload

[Install]
WantedBy=sysinit.target
```




But I also notice this part in the `etc/init.d/apparmor` script, which seems to refuse running in a container, but not sure if that's relevant in any way;

```bash
case "$1" in
	start)
		if [ -x /usr/bin/systemd-detect-virt ] && \
		   systemd-detect-virt --quiet --container && \
		   ! is_container_with_internal_policy; then
			log_daemon_msg "Not starting AppArmor in container"
			log_end_msg 0
			exit 0
		fi
		log_daemon_msg "Starting AppArmor profiles"
		securityfs
		load_configured_profiles
		rc=$?
		log_end_msg "$rc"
		;;
```

```bash
/usr/bin/systemd-detect-virt --help
systemd-detect-virt [OPTIONS...]

Detect execution in a virtualized environment.

  -h --help             Show this help
     --version          Show package version
  -c --container        Only detect whether we are run in a container
  -v --vm               Only detect whether we are run in a VM
  -r --chroot           Detect whether we are run in a chroot() environment
     --private-users    Only detect whether we are running in a user namespace
  -q --quiet            Don't output anything, just set return value
```

thaJeztah commented:
Right, that error comes from systemd (`check_wait_response`);
https://github.com/systemd/systemd/blob/ed272a9ff59a26beedaab508dd3c9d631de67165/src/shared/bus-wait-for-jobs.c#L244-L245

```c
else if (streq(d->result, "assert"))
    log_error("Assertion failed on job for %s.", strna(d->name));
else if (streq(d->result, "unsupported"))
```

thaJeztah commented:
Interesting; so it looks like this actually failed on bullseye as well, but _silently_!

On bookworm, AppArmor fails to start inside the container, which can be seen at startup of the dev-container:

    Created symlink /etc/systemd/system/systemd-firstboot.service → /dev/null.
    Created symlink /etc/systemd/system/systemd-udevd.service → /dev/null.
    Created symlink /etc/systemd/system/multi-user.target.wants/docker-entrypoint.service → /etc/systemd/system/docker-entrypoint.service.
    hack/dind-systemd: starting /lib/systemd/systemd --show-status=false --unit=docker-entrypoint.target
    systemd 252.17-1~deb12u1 running in system mode (+PAM +AUDIT +SELINUX +APPARMOR +IMA +SMACK +SECCOMP +GCRYPT -GNUTLS +OPENSSL +ACL +BLKID +CURL +ELFUTILS +FIDO2 +IDN2 -IDN +IPTC +KMOD +LIBCRYPTSETUP +LIBFDISK +PCRE2 -PWQUALITY +P11KIT +QRENCODE +TPM2 +BZIP2 +LZ4 +XZ +ZLIB +ZSTD -BPF_FRAMEWORK -XKBCOMMON +UTMP +SYSVINIT default-hierarchy=unified)
    Detected virtualization docker.
    Detected architecture x86-64.
    modprobe@configfs.service: Deactivated successfully.
    modprobe@dm_mod.service: Deactivated successfully.
    modprobe@drm.service: Deactivated successfully.
    modprobe@efi_pstore.service: Deactivated successfully.
    modprobe@fuse.service: Deactivated successfully.
    modprobe@loop.service: Deactivated successfully.
    apparmor.service: Starting requested but asserts failed.
    proc-sys-fs-binfmt_misc.automount: Got automount request for /proc/sys/fs/binfmt_misc, triggered by 49 (systemd-binfmt)
    + source /etc/docker-entrypoint-cmd
    ++ hack/make.sh dynbinary test-integration

When checking "aa-status", an error was printed that the filesystem was
not mounted:

    aa-status
    apparmor filesystem is not mounted.
    apparmor module is loaded.

Checking if "local-fs.target" was loaded, that seemed to be the case;

    systemctl status local-fs.target
    ● local-fs.target - Local File Systems
         Loaded: loaded (/lib/systemd/system/local-fs.target; static)
         Active: active since Mon 2023-11-27 10:48:38 UTC; 18s ago
           Docs: man:systemd.special(7)


However, **on the host**, "/sys/kernel/security" has a mount, which was not
present inside the container:

    mount | grep securityfs
    securityfs on /sys/kernel/security type securityfs (rw,nosuid,nodev,noexec,relatime)


Interestingly, on `debian:bullseye`, this was not the case either; no
`securityfs` mount was present inside the container, and apparmor actually
failed to start, but succeeded silently:

    mount | grep securityfs
    systemctl start apparmor
    systemctl status apparmor
    ● apparmor.service - Load AppArmor profiles
         Loaded: loaded (/lib/systemd/system/apparmor.service; enabled; vendor preset: enabled)
         Active: active (exited) since Mon 2023-11-27 11:59:09 UTC; 44s ago
           Docs: man:apparmor(7)
                 https://gitlab.com/apparmor/apparmor/wikis/home/
        Process: 43 ExecStart=/lib/apparmor/apparmor.systemd reload (code=exited, status=0/SUCCESS)
       Main PID: 43 (code=exited, status=0/SUCCESS)
            CPU: 10ms
    
    Nov 27 11:59:09 9519f89cade1 apparmor.systemd[43]: Not starting AppArmor in container

Same, using the `/etc/init.d/apparmor` script:

    /etc/init.d/apparmor start
    Starting apparmor (via systemctl): apparmor.service.
    echo $?
    0

And apparmor was not actually active:

    aa-status
    apparmor module is loaded.
    apparmor filesystem is not mounted.

    aa-enabled
    Maybe - policy interface not available.

thaJeztah commented:
After further investigating, I found that the non-systemd dind script had a mount for AppArmor, which was added in 31638ab2ad2a5380d447780f05f7aa078c9421f5 (https://github.com/moby/moby/pull/2497)

The systemd variant was missing this mount, which may have gone unnoticed because `debian:bullseye` was silently ignoring this when starting the apparmor service.

thaJeztah commented:
NIce! This is green now, so I can rebase this one, once https://github.com/moby/moby/pull/46852 is merged

thaJeztah commented:
@neersighted @tianon this one should be ready now (was been before I rebased)

cpuguy83 commented on integration-cli/docker_cli_daemon_test.go:
```diff
@@ -798,7 +798,7 @@ func (s *DockerDaemonSuite) TestDaemonICCPing(c *testing.T) {
 	result := icmd.RunCommand("iptables", "-nvL", "FORWARD")
 	result.Assert(c, icmd.Success)
-	regex := fmt.Sprintf("DROP.*all.*%s.*%s", bridgeName, bridgeName)
+	regex := fmt.Sprintf("DROP.*0.*%s.*%s", bridgeName, bridgeName)
```
I have almost no confidence in the beforehand regex (`DROP.*` is pretty loose) , now I have less :(
Are we sure this is correct? Ubuntu still has `all`.

If there really are differences here, we should handle this better as the test seems tied to the distro release.

thaJeztah replied:
Yeah, not super-stoked with the test as well. Looking as well why on earth we used `-v` (verbose); it looks like without that, it's already a lot more parseable;

```bash
iptables -vL FORWARD
Chain FORWARD (policy ACCEPT 0 packets, 0 bytes)
pkts bytes target     prot opt in     out     source               destination
0     0    DOCKER-USER  all  --  any    any     anywhere             anywhere
0     0    DOCKER-ISOLATION-STAGE-1  all  --  any    any     anywhere             anywhere
0     0    ACCEPT     all  --  any    ext-bridge6  anywhere             anywhere             ctstate RELATED,ESTABLISHED
0     0    DOCKER     all  --  any    ext-bridge6  anywhere             anywhere
0     0    ACCEPT     all  --  ext-bridge6 !ext-bridge6  anywhere             anywhere
0     0    DROP       all  --  ext-bridge6 ext-bridge6  anywhere             anywhere
```

```bash
iptables -L FORWARD
Chain FORWARD (policy ACCEPT)
target     prot opt source               destination
DOCKER-USER  all  --  anywhere             anywhere
DOCKER-ISOLATION-STAGE-1  all  --  anywhere             anywhere
ACCEPT     all  --  anywhere             anywhere             ctstate RELATED,ESTABLISHED
DOCKER     all  --  anywhere             anywhere
ACCEPT     all  --  anywhere             anywhere
DROP       all  --  anywhere             anywhere
```

thaJeztah replied:
🙈 nevermind; non-verbose excludes the `in` and `out` columns

akerouanton replied:
> Are we sure this is correct? Ubuntu still has all.

Yeah, iptables' `-n` means:

> Numeric  output.   IP  addresses  and port numbers will be printed in numeric format.

There was a bug in older iptables versions (bullseye is shipping iptables v1.8.7; and bookworm is shipping iptables v1.8.9) that made it not respect `-n`. This was fixed in https://git.netfilter.org/iptables/commit/?id=da8ecc62dd765b15df84c3aa6b83dcb7a81d4ffa.

But I agree this regex is not super great.

thaJeztah replied:
- opened https://github.com/moby/moby/pull/46859 with an alternative

tianon commented on Dockerfile:
```diff
@@ -40,14 +39,12 @@ FROM --platform=$BUILDPLATFORM ${GOLANG_IMAGE} AS base
 ARG APT_MIRROR
-RUN test -n "$APT_MIRROR" && sed -ri "s#(httpredir|deb|security).debian.org#${APT_MIRROR}#g" /etc/apt/sources.list || true
-ARG DEBIAN_FRONTEND
+RUN test -n "$APT_MIRROR" && sed -ri "s#(httpredir|deb|security).debian.org#${APT_MIRROR}#g" /etc/apt/sources.list.d/debian.sources || true
```
Still of the opinion that we should just drop the `APT_MIRROR` code, but non-blocking. :innocent:

thaJeztah replied:
Yeah, not strongly against, mostly curious if there's still users for it (I somewhat assume people behind the great firewall would already have things in place for such things though)

I'll keep that as a follow-up PR, also to make such a change more visible (instead of implicit change in an update)

thaJeztah replied:
- opened https://github.com/moby/moby/pull/46884 for this

tianon commented on Dockerfile:
```diff
@@ -559,7 +564,8 @@ RUN --mount=type=cache,sharing=locked,id=moby-dev-aptlib,target=/var/lib/apt \
             libsystemd-dev \
-            libudev-dev
+            libudev-dev \
+            yamllint=${YAMLLINT_VERSION}\*
```
This is going to cause problems over time -- the Debian archive isn't going to keep old versions, so instead of pinning to a specific version forever, this is going to ensure that we eventually just fail to build (although likely not until the next Debian release unless something _really_ exceptional happens, so I guess we're mostly probably OK for now?)

thaJeztah replied:
Good point. I think we're pinning this one so that we didn't get random linter updates causing CI to fail, but we should definitely keep it a bit up to date periodically.

Still looking as well if there's a good alternative not involving python. I think we're mainly interested in well-formed-ness (need to check again where we're all using it)

thaJeztah replied:
- opened https://github.com/moby/moby/pull/46884 for this

tianon commented on cli/winresources/README.md:
````diff
@@ -1,7 +1,7 @@
 
 ```console
-$ docker run --rm -it -v "$(pwd):/winresources" debian:bullseye bash
+$ docker run --rm -it -v "$(pwd):/winresources" debian:bookworm bash
````
Should we update this one to `-slim` too while we're at it, for consistency? (I think we're now using `-slim` pretty much everywhere else :sweat_smile:)

thaJeztah replied:
Ah, yeah, I'm all for "slim" where we can!

thaJeztah replied:
updated; we should probably look at having a dockerfile for this, and to make sure it's verified as part of the "generated files", which made me recall we still want to move this package as well;

- https://github.com/moby/moby/issues/44641

tianon reviewed (approved):
A couple minor comments, none of which are very important or blocking :bow:

thaJeztah commented on Dockerfile:
```diff
@@ -529,9 +535,6 @@ RUN --mount=type=cache,sharing=locked,id=moby-dev-aptlib,target=/var/lib/apt \
             net-tools \
             patch \
             pigz \
-            python3-pip \
-            python3-setuptools \
-            python3-wheel \
```
looks like yamllint was happy to install without these, but maybe they were needed for cross-compile, so let's see if CI is happy.

crazy-max replied:
I think this was needed for `pip3 install`, to pin to a specific version of yamlint

crazy-max commented:
I see something has been left behind https://github.com/moby/moby/blob/7dc5f9d0b12720ba85064cbe2cbbc41e7e6f69b5/Dockerfile#L49 :eyes:

Should be

```
https://download.opensuse.org/repositories/devel:/tools:/criu/Debian_12/
```

thaJeztah commented:
Ha! Yes, we were just discussing that earlier.
